matrix_sdk/src/client.rs

```diff
@@ -679,8 +714,205 @@ impl Client {
 
         let request = assign!(
             login::Request::new(
-                login::UserInfo::MatrixId(user),
-                login::LoginInfo::Password { password },
+                login::LoginInfo::Password { identifier: login::UserIdentifier::MatrixId(user), password },
+            ), {
+                device_id: device_id.map(|d| d.into()),
+                initial_device_display_name,
+            }
+        );
+
+        let response = self.send(request, None).await?;
+        self.base_client.receive_login_response(&response).await?;
+
+        Ok(response)
+    }
+
+    /// Login to the server via Single Sign-On.
+    ///
+    /// This takes care of the whole SSO flow:
+    ///   * Spawn a local http server
+    ///   * Open the SSO login URL in the default web browser
+    ///   * Wait for the local http server to get the loginToken
+    ///   * Call [`login_with_token`]
+    ///
+    /// If cancellation is needed the method should be wrapped in a cancelable task
+    ///
+    /// If you need more control over the SSO login process, you should use
+    /// [`get_sso_login_url`] and [`login_with_token`] directly.
+    ///
+    /// This should only be used for the first login, note that if the device id
+    /// isn't provided a new device will be created.
+    ///
+    /// If this isn't the first login, a device id should be provided to restore
+    /// the correct stores.
+    ///
+    /// The [`restore_login`] method should be used to restore a
+    /// logged in client after the first login.
+    ///
+    /// # Arguments
+    ///
+    /// * `server_url` - The local URL the server is going to try to bind to, e.g.
+    ///     `http://127.0.0.1:3030`.
+    ///
+    /// * `device_id` - A unique id that will be associated with this session. If
+    ///     not given the homeserver will create one. Can be an existing device_id
+    ///     from a previous login call. Note that this should be provided only
+    ///     if the client also holds the encryption keys for this device.
+    ///
+    /// * `initial_device_display_name` - A public display name that will be
+    ///     associated with the device_id. Only necessary the first time you
+    ///     login with this device_id. It can be changed later.
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use std::convert::TryFrom;
+    /// # use matrix_sdk::Client;
+    /// # use matrix_sdk::identifiers::DeviceId;
+    /// # use matrix_sdk_common::assign;
+    /// # use futures::executor::block_on;
+    /// # use url::Url;
+    /// # let homeserver = Url::parse("https://example.com").unwrap();
+    /// # block_on(async {
+    /// let client = Client::new(homeserver).unwrap();
+    ///
+    /// let response = client
+    ///     .login_with_sso(None, Some("My app"))
```

Comment:
This example doesn't match reality. The tests fail for this, but CI passes since the feature is disabled by default, we should probably add another CI task that tests this feature.

Reply:
Yeah, sorry about that. I forgot to update the example after implementing the function.

Reply:
Should I take care of the new CI task here or in another PR? I'm also not sure what command to use because locally when I launch `cargo test --features docs`, it applies the regular test but not the docs test of the `login_with_sso` method.